### PR TITLE
Add -WithIndex classes

### PR DESCRIPTION
**Pull request by mbid**
Add -WithIndex classes

This adds `FunctorWithIndex`, `FoldableWithIndex` and
`TraversableWithIndex` and some combinators. Implementations are mostly
copied from the versions without index with minor adaptions. Similarly
for the documentation.

I think the corresponding haskell package does not state all laws for
these type classes, so there are TODO's in the doc strings of the three
classes.

`StateL` and `StateR` in TraversableWithIndex.purs are exact copies of
the versions in Traversable.purs because they are not exported there.
Maybe it would be better to merge TraversableWithIndex.purs into
Traversable.purs and similarly for Foldable. The Foldable and Traverse
instances of Array could then be along the lines of
```purescript
traverse f = itraverse (const f)`
```
so that the foreign js imports for the unindexed version could be
removed.

`FunctorWithIndex`'s `imap` collides with the `imap` of
purescript-invariant. Haskell calls the latter `invmap`. I don't know
whether that's a problem.

This does not yet include the `-1WithIndex` classes, i.e.
generalizations of `Foldable1` and `Traversable1` in `Semigroup/`. I
guess it would make sense to provide them here as well.

If this gets merged we should also implement instances in some other packages. purescript-lists and purescript-maps comes to mind. Anywhere else?

### CODE REVIEW

File: src/Data/FunctorWithIndex.purs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+-- | Instances must satisfy a modified form of the `Functor` laws: 
+-- | 
+-- | - Identity: `imap (\_ a -> a) = id`
+-- | - Composition: `imap f . imap g = imap (\i -> f i <<< g i)`
+-- |
+-- | TODO: Don't we also want
+-- | - `imap (const f) = map f`
+-- | though?
+-- |
+-- | TODO: `imap` collides with `Invariant`s `imap`.
```
Comment (garyb) [inline src/Data/FunctorWithIndex.purs]: Maybe we should either use `WithIndex` suffixes on all of these functions, or prefix with `ix` as an alternative to `i` prefixes?

I wouldn't be too worried about overlapping with `Invariant`'s map either though, there's always qualification. And `Invariant` is one of the less frequently used classes.

**Comment by paf31** [comment]
Thanks!

I think we can remove the FFI requirement here by defining instances in terms of regular `Foldable` and `Traversable` and `Array.mapWithIndex`, right? (at the expense of a little performance admittedly)

What do you think?

**Comment by paf31** [comment]
For the `StateX` thing, I think we could make an internal module.

**Comment by mbid** [comment]
Yeah, I agree that the duplication of the FFI implementations is not nice. I guess we can do two things:

* What you said: Use `imap`, i.e. `itraverse f = sequence <<< imap f`. The `FoldableWithIndex` functions can be implemented either using `Const` and this `itraverse`, or `imap Tuple` followed by the folds without index. Unfortunately, we'd need to implement `Const` or `Tuple` here, because both packages already depend on foldable-traversable.
* Implement the variants without index in terms of the indexed variants, e.g. `foldMap = ifoldMap <<< const`. This would require defining the -WithIndex classes in the same file together with the variants without index (e.g. `FoldableWithIndex` in Foldable.purs), or moving the FFI implementations to a separate internal module. Maybe `foldMap` would be a bit slower because of the additional argument passed, but `ifoldMap` would only iterate the array once instead of twice, and use no additional memory (similarly for the other functions).

I'd prefer the second version slightly, but let me know which one you'd find better.

File: src/Data/FunctorWithIndex.purs (hunk 1 above)
Comment (mbid) [inline src/Data/FunctorWithIndex.purs]: I'd prefer `ix-` over `-WithIndex`. Consistency with the haskell functions (so `i-`) is also nice I guess. Your call.

**Comment by mbid** [comment]
I've just done [some benchmarks](https://github.com/mbid/purescript-ifoldmap-benchmark) for the second approach. Looks like this would make `foldMap` around 25% slower than it currently is.

**Comment by mbid** [comment]
Ok, so I've removed the foreign implementations where possible. This means that `ifoldr` and `ifoldl` are only 1/2 - 1/3 as fast as they could to be for trivial iteration (e.g. just adding integers), but that's good for me. Given that the versions without indices are used far more often, I don't think it would've been a good tradeoff to take a performance hit for `foldr` instead of `ifoldr`.

Regarding the documentation, I've documented some laws that enforce that `map f = imap (const f)` and analogous equations for `Foldable` and `Traversable`.  
Given that the complete `Foldable` and `Traversable` laws are not in the docs either, I found it pointless to document the modified versions for the `-WithIndex` versions. We should maybe improve this in a separate issue.

I guess I'll add some tests now. Can we break up the single test file into one for each class? As far as I can tell, there is no interdependence.

**Comment by mbid** [comment]
Ok so I've just gone ahead and added some tests.

File: src/Data/FunctorWithIndex.purs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+-- | ```purescript
+-- | imap (\_ a -> a) = id
+-- | imap f . imap g = imap (\i -> f i <<< g i)
+-- | ```
+-- | and be compatible with the `Functor` instance
+-- | ```purescript
+-- | map f = imap (const f)
+-- | ```
+class Functor f <= FunctorWithIndex i f | f -> i where
+  imap :: forall a b. (i -> a -> b) -> f a -> f b
```
Comment (paf31) [inline src/Data/FunctorWithIndex.purs]: This is called `mapWithIndex` [elsewhere](https://pursuit.purescript.org/packages/purescript-arrays/4.1.1/docs/Data.Array#v:mapWithIndex) actually.

File: src/Data/FunctorWithIndex.purs (hunk 1 above)
Comment (paf31) [inline src/Data/FunctorWithIndex.purs]: I'd prefer `mapWithIndex` since we use that convention elsewhere.

**Comment by paf31** [comment]
@garyb Any more thoughts on this? Other than naming, I think this looks great.

**Comment by mbid** [comment]
Are you suggesting renaming only `imap` to `mapWithIndex`, or also `ifoldMap` to `foldMapWithIndex` and so on? I'd find it a little inconsistent if we only did the first thing.

**Comment by paf31** [comment]
I'm suggesting renaming them all.

**Comment by mbid** [comment]
Ok, done.

File: src/Data/Bifoldable.purs [hunk 3]
```diff
@@ -26,131 +26,131 @@ import Data.Bifunctor.Wrap (Wrap(..))
...
+-- | - `bfoldMapWithIndexDefaultL`
 -- |
 -- | Note: some combinations of the default implementations are unsafe to
 -- | use together - causing a non-terminating mutually recursive cycle.
 -- | These combinations are documented per function.
 class Bifoldable p where
-  bifoldr :: forall a b c. (a -> c -> c) -> (b -> c -> c) -> c -> p a b -> c
-  bifoldl :: forall a b c. (c -> a -> c) -> (c -> b -> c) -> c -> p a b -> c
-  bifoldMap :: forall m a b. Monoid m => (a -> m) -> (b -> m) -> p a b -> m
+  bfoldrWithIndex :: forall a b c. (a -> c -> c) -> (b -> c -> c) -> c -> p a b -> c
```
Comment (paf31) [inline src/Data/Bifoldable.purs]: I'm confused why these changed.

Comment (mbid) [inline src/Data/Bifoldable.purs]: Uh yeah, and rightly so.

**Comment by paf31** [comment]
Great work, thank you!

I'll make a minor release for this since nothing seems to be breaking here.

**Pull request merged**